src/librbd/object_map/DiffRequest.cc

```diff
@@ -24,18 +24,16 @@ template <typename I>
 void DiffRequest<I>::send() {
   std::shared_lock image_locker{m_image_ctx->image_lock};
 
-  m_diff_from_start = (m_snap_id_start == 0);
-  if (m_snap_id_start == 0) {
-    if (!m_image_ctx->snap_info.empty()) {
-      m_snap_id_start = m_image_ctx->snap_info.begin()->first;
-    } else {
-      m_snap_id_start = CEPH_NOSNAP;
-    }
-  }
-
   m_object_diff_state->clear();
-  m_current_snap_id = m_snap_id_start;
-  m_next_snap_id = m_snap_id_end;
+
+  // collect all the snap ids in the provided range (inclusive)
+  m_snap_ids.insert(m_snap_id_start);
+  auto snap_info_it = m_image_ctx->snap_info.upper_bound(m_snap_id_start);
+  auto snap_info_it_end = m_image_ctx->snap_info.lower_bound(m_snap_id_end);
+  for (; snap_info_it != snap_info_it_end; ++snap_info_it) {
```

Comment:
nit: Wouldn't it make sense to have `ceph_assert(m_snap_id_start < m_snap_id_end)` before doing this?

Reply:
Added above (just not as an assert)